src/lib/pickles/verify.ml

```diff
@@ -148,31 +148,49 @@ let verify_heterogenous (ts : Instance.t list) =
           in
           (absorb sponge, squeeze)
         in
-        ( absorb evals.ft_eval1 ;
-          let xs = Plonk_types.Evals.to_absorption_sequence evals.evals.evals in
-          let x1, x2 = evals.evals.public_input in
-          absorb x1 ;
-          absorb x2 ;
-          List.iter xs ~f:(fun (x1, x2) ->
-              Array.iter ~f:absorb x1 ; Array.iter ~f:absorb x2 ) ) ;
+        let old_bulletproof_challenges =
+          Vector.map ~f:Ipa.Step.compute_challenges
+            statement.pass_through.old_bulletproof_challenges
+        in
+        let sg_evals1, sg_evals2 =
+          let challenge_polys =
+            let vec =
+              Vector.map
+                ~f:(fun chals ->
+                  unstage (Wrap.challenge_polynomial (Vector.to_array chals)) )
+                old_bulletproof_challenges
+            in
+            fun pt -> Vector.map vec ~f:(fun f -> f pt)
+          in
+          (challenge_polys zeta, challenge_polys zetaw)
+        in
+        (let sg_eval_digest =
+           let open Tick_field_sponge.Field in
+           let sponge = create Tick_field_sponge.params in
+           Vector.iter2 sg_evals1 sg_evals2 ~f:(fun sg_eval1 sg_eval2 ->
+               absorb sponge sg_eval1 ; absorb sponge sg_eval2 ) ;
+           squeeze sponge
+         in
```

Comment:
I have a hard time understanding why we create an internal sponge at this point as an optimization, might be good to write an explainer on that somewhere (hackmd, minabook, etc.)